app/services/samples/metadata/update_service.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Samples
+  module Metadata
+    # Service used to Update Samples::Metadata
+    class UpdateService < BaseService
+      SampleMetadataUpdateError = Class.new(StandardError)
+      attr_accessor :sample, :metadata, :analysis_id
+
+      def initialize(project, sample, user = nil, params = {})
+        super(user, params)
+        @project = project
+        @sample = sample
+        @metadata = params['metadata']
+        @analysis_id = params['analysis_id']
+      end
+
+      def execute # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength
+        authorize! sample.project, to: :update_sample?
+
+        if @project.id != @sample.project.id
+          raise SampleMetadataUpdateError,
+                I18n.t('services.samples.metadata.sample_does_not_belong_to_project', sample_name: @sample.name,
+                                                                                      project_name: @project.name)
+        end
+
+        if @metadata
+          # Without transforming keys, issues with overwritting can occur and multiples of the same key can appear
+          @metadata = @metadata.transform_keys(&:to_s)
+          @metadata.each do |key, value|
+            if value.blank?
+              if @sample['metadata'].key?(key)
+                @sample['metadata'].delete(key) && @sample['metadata_provenance'].delete(key)
+              end
+            else
+              update_metadata(key, value, current_user)
+            end
+          end
+        else
+          raise SampleMetadataUpdateError,
+                I18n.t('services.samples.metadata.empty_metadata', sample_name: @sample.name)
+        end
+
+        @sample.update(id: @sample.id)
+      rescue Samples::Metadata::UpdateService::SampleMetadataUpdateError => e
+        @sample.errors.add(:base, e.message)
+      end
+
+      private
+
+      def update_metadata(key, value, user)
+        if @sample['metadata_provenance'].key?(key)
```

Comment:
In this method we should probably return which key -> value wasn't updated due to the previous value being set by an analysis and now a user is trying to update it. Then in the `execute` method above within the .each loop keep track of the keys that are returned from the `update_metadata` method. We can then attach an error message which lists the keys that weren't updated to `samples.errors -> base`.

Reply:
Updated in 3b73978b5eaf0aa163e17bb5034f7b9b62c8bf9b
Message updated in 0e24012480f59df2a7803e01baa1d8a8f8192615

Reply:
I also have both the metadata field and value in the message, so it will read: `Metadata fields 'key1: value1, key2: value2, etc.' of sample {sample_name} cannot be overwritten by a user as it was previously updated with an analysis. `

In terms of error msg readability, I'm not sure value is necessary here and the metadata field is sufficient for letting the user know it didn't update. Thoughts?